### PR TITLE
Мухутдинов Дмитрий, 2538: Lesson10

### PR DESCRIPTION
**Pull request by flyingleafe**
Turns out what a good Android app should take much more time than any other educational task.
Not everything is done, and things which are done aren't perfect.

### CODE REVIEW

File: app/src/main/java/ru/ifmo/md/lesson8/AddCityActivity.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+import ru.ifmo.md.lesson8.service.ForecastService;
+import ru.ifmo.md.lesson8.service.Throttler;
+
+
+public class AddCityActivity extends ActionBarActivity implements TextWatcher, AsyncResponseReceiver<ForecastParser.LocationResult> {
+
+    EditText searchField;
+    TextView suggestion;
+    ForecastParser.LocationResult currentResult;
+    Throttler mThrottler = new Throttler();
```
Comment (thevery) [inline app/src/main/java/ru/ifmo/md/lesson8/AddCityActivity.java:29]: inconsistent naming

File: app/src/main/java/ru/ifmo/md/lesson8/AddCityActivity.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        if(s.length() > 0) {
+            final String name = s.toString();
+            final CitySearchTask task = new CitySearchTask(this);
+            Throttler.Callback foo = new Throttler.Callback() {
+                @Override
+                public void call() {
+                    task.execute(name);
+                }
+            };
+            mThrottler.throttle(foo, 500);
```
Comment (thevery) [inline app/src/main/java/ru/ifmo/md/lesson8/AddCityActivity.java:56]: what is the idea of using throttler here?

File: app/src/main/java/ru/ifmo/md/lesson8/AddCityActivity.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+        }
+    }
+
+    @Override
+    public void afterTextChanged(Editable s) {}
+
+    @Override
+    public void processFinish(ForecastParser.LocationResult result) {
+        if(result == null) {
+            suggestion.setText("An error occured, sorry :(");
```
Comment (thevery) [inline app/src/main/java/ru/ifmo/md/lesson8/AddCityActivity.java:68]: `strings.xml`!

File: app/src/main/java/ru/ifmo/md/lesson8/CitiesListFragment.java [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+        inf.addAction(ForecastService.ACTION_GET_LOCATION_WOEID);
+        inf.addAction(ForecastService.ACTION_UPDATE_CITIES_LIST);
+        LocalBroadcastManager.getInstance(getActivity()).registerReceiver(mBroadcastReceiver, inf);
+    }
+
+    @Override
+    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
+        View v = super.onCreateView(inflater, container, savedInstanceState);
+        ViewGroup parent = (ViewGroup) inflater.inflate(R.layout.nav_drawer, container, false);
+        parent.addView(v, 2);
```
Comment (thevery) [inline app/src/main/java/ru/ifmo/md/lesson8/CitiesListFragment.java:132]: looks like a hack

File: app/src/main/java/ru/ifmo/md/lesson8/CitiesListFragment.java [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+        }
+
+        mActivatedPosition = position;
+    }
+
+    @Override
+    public Loader<Cursor> onCreateLoader(int id, Bundle args) {
+        return new CursorLoader(getActivity(),
+                WeatherContentProvider.CITIES_CONTENT_URL,
+                new String[] {"*"},
```
Comment (thevery) [inline app/src/main/java/ru/ifmo/md/lesson8/CitiesListFragment.java:257]: do you really need this?

File: app/src/main/java/ru/ifmo/md/lesson8/service/CitySearchTask.java [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    protected ForecastParser.LocationResult doInBackground(String... params) {
+        String name = params[0];
+        String searchUrl = CITY_SEARCH_URI.replace("{{ cityName }}", name);
+        try {
+            InputStream response = new URL(searchUrl).openStream();
+            return new ForecastParser().parseLocation(response);
+        } catch (IOException | SAXException e) {
+            e.printStackTrace();
+            return null;
+        }
```
Comment (thevery) [inline app/src/main/java/ru/ifmo/md/lesson8/service/CitySearchTask.java:35]: `close` in finally?

File: app/src/main/java/ru/ifmo/md/lesson8/service/AlarmReceiver.java [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+
+public class AlarmReceiver extends BroadcastReceiver {
+    public AlarmReceiver() {
+    }
+
+    @Override
+    public void onReceive(Context context, Intent intent) {
+        Intent newIntent = new Intent(context, ForecastService.class);
+        newIntent.setAction(ForecastService.ACTION_UPDATE_ALL);
+        context.startService(newIntent);
```
Comment (thevery) [inline app/src/main/java/ru/ifmo/md/lesson8/service/AlarmReceiver.java:15]: why not service directly?

**Comment by thevery** [comment]
- no app icon
- modal progress on first load is BAD
- extra click for forecast is redundant
- "add this city" with empty input crashes app (`Caused by: java.lang.NullPointerException
  E/AndroidRuntime(13182):    at ru.ifmo.md.lesson8.AddCityActivity.addCity(AddCityActivity.java:77)`)
- "current location" doesn't always work

**Comment by thevery** [comment]
ok
24-4=20
(code and app are pretty good, but lacks for minor, but important details)

File: app/src/main/java/ru/ifmo/md/lesson8/AddCityActivity.java (hunk 2 above)
Comment (flyingleafe) [inline app/src/main/java/ru/ifmo/md/lesson8/AddCityActivity.java:56]: The idea is not to send a request until user stops typing. Throttler waits given number of ms before starting the task, and this timeout is resetted if `throttle()` is called again before expiration. As a result, nothing happens when text in the field changes frequently and a request is sent only half a second after text stopped changing.

Comment (thevery) [inline app/src/main/java/ru/ifmo/md/lesson8/AddCityActivity.java:56]: I don't think that extra delay is a good idea

**Pull request closed without merging**